### ISSUE TITLE
TSX not highlighting parts

### ISSUE THREAD

**Issue by louisrli** [rouge-ruby/rouge#1363]
**Name of the lexer**
tsx
**Code sample**
A sample of the code that produces the bug.

```
const withAuthRedirect = (route: string, redirectIfAuthed: boolean) => <P,>(
  Page: NextComponentType<NextPageContext, {}, P>
) => {
  return class extends React.Component<P> {
    static async getInitialProps(ctx: NextPageContext) {
      const shouldContinue = await redirectBasedOnLogin(
        ctx,
        route,
        redirectIfAuthed
      );
      // Only continue if we're logged in. Otherwise, it might cause an
      // unnecessary call to a downstream getInitialProps that requires
      // authentication.
      if (!shouldContinue) {
        return {};
      }
      if (Page.getInitialProps) {
        return Page.getInitialProps(ctx);
      }
    }

    render() {
      return <Page {...this.props} />;
    }
  };
};

/**
 * HOC that redirects to login page if the user is not logged in.
 */
export const withLoginRedirect = withAuthRedirect('/login', false);

/**
 * HOC that redirects to the dashboard if the user is logged in.
 */
export const withDashboardRedirect = withAuthRedirect('/dashboard', true);
```

Last two exports don't get highlighted.
**Screenshots**
![image](https://user-images.githubusercontent.com/1303401/69002832-1a6ece80-0929-11ea-9c4a-aac976e7c47f.png)

**Additional context**
Deleting the whole variable `withAuthRedirect` in the example above fixes the issue.

**Comment by pyrmont** [comment]
A [live example](http://rouge.jneen.net/v3.17.0/tsx/Y29uc3Qgd2l0aEF1dGhSZWRpcmVjdCA9IChyb3V0ZTogc3RyaW5nLCByZWRpcmVjdElmQXV0aGVkOiBib29sZWFuKSA9PiA8UCw-KAogIFBhZ2U6IE5leHRDb21wb25lbnRUeXBlPE5leHRQYWdlQ29udGV4dCwge30sIFA-CikgPT4gewogIHJldHVybiBjbGFzcyBleHRlbmRzIFJlYWN0LkNvbXBvbmVudDxQPiB7CiAgICBzdGF0aWMgYXN5bmMgZ2V0SW5pdGlhbFByb3BzKGN0eDogTmV4dFBhZ2VDb250ZXh0KSB7CiAgICAgIGNvbnN0IHNob3VsZENvbnRpbnVlID0gYXdhaXQgcmVkaXJlY3RCYXNlZE9uTG9naW4oCiAgICAgICAgY3R4LAogICAgICAgIHJvdXRlLAogICAgICAgIHJlZGlyZWN0SWZBdXRoZWQKICAgICAgKTsKICAgICAgLy8gT25seSBjb250aW51ZSBpZiB3ZSdyZSBsb2dnZWQgaW4uIE90aGVyd2lzZSwgaXQgbWlnaHQgY2F1c2UgYW4KICAgICAgLy8gdW5uZWNlc3NhcnkgY2FsbCB0byBhIGRvd25zdHJlYW0gZ2V0SW5pdGlhbFByb3BzIHRoYXQgcmVxdWlyZXMKICAgICAgLy8gYXV0aGVudGljYXRpb24uCiAgICAgIGlmICghc2hvdWxkQ29udGludWUpIHsKICAgICAgICByZXR1cm4ge307CiAgICAgIH0KICAgICAgaWYgKFBhZ2UuZ2V0SW5pdGlhbFByb3BzKSB7CiAgICAgICAgcmV0dXJuIFBhZ2UuZ2V0SW5pdGlhbFByb3BzKGN0eCk7CiAgICAgIH0KICAgIH0KCiAgICByZW5kZXIoKSB7CiAgICAgIHJldHVybiA8UGFnZSB7Li4udGhpcy5wcm9wc30gLz47CiAgICB9CiAgfTsKfTsKCi8qKgogKiBIT0MgdGhhdCByZWRpcmVjdHMgdG8gbG9naW4gcGFnZSBpZiB0aGUgdXNlciBpcyBub3QgbG9nZ2VkIGluLgogKi8KZXhwb3J0IGNvbnN0IHdpdGhMb2dpblJlZGlyZWN0ID0gd2l0aEF1dGhSZWRpcmVjdCgnL2xvZ2luJywgZmFsc2UpOwoKLyoqCiAqIEhPQyB0aGF0IHJlZGlyZWN0cyB0byB0aGUgZGFzaGJvYXJkIGlmIHRoZSB1c2VyIGlzIGxvZ2dlZCBpbi4KICovCmV4cG9ydCBjb25zdCB3aXRoRGFzaGJvYXJkUmVkaXJlY3QgPSB3aXRoQXV0aFJlZGlyZWN0KCcvZGFzaGJvYXJkJywgdHJ1ZSk7) is available on the Dingus.

The problem appears to be caused by lexing errors that are occurring with the `withAuthRedirect` function. Specifically:

1. `=> <P,>`: The `>` coming immediately after the comma causes an error.
2. `<NextComponentType<NextPageContext, {}, P>`: The commas are lexing as errors.
3. `<P,>(...)`: The closing parenthesis is lexing as an error.
4. `<Page {...this.props} />`: The use of braces is lexing as an error.

@louisrli I assume that these errors are incorrect? I'm not familiar with TSX; would you be able to link to documentation explaining how the syntax given above? That would help in correcting Rouge's lexer.

**Comment by louisrli** [comment]
So the lexer I'm using is `tsx` which is slightly different from `ts` (or it should be). TypeScript is basically JavaScript with types, while TSX is derived from Facebook's React framework and essentially allows embedding HTML into the JS as values (which complicates the syntax). At least rouge seems to have a special lexer for TSX, so I'm not sure if the live example you sent is using the exact same lexer.

For 1 specifically, I spent a while searching the [spec](https://github.com/Microsoft/TypeScript/blob/730f18955dc17068be33691f0fb0e0285ebbf9f5/doc/spec.md#3.11.7) and couldn't find something official on the trailing comma...but it is TSX specific

However, Typescript sometimes is used with Facebook's react library to form this weird mesh of Typescript + HTML. I did find why the trailing comma was there, under ["Arrow functions" of this article](https://wanago.io/2020/02/17/typescript-generics-discussing-naming-conventions/)

```
The trailing comma in  <ObjectType, > is added due to contraints of the  .tsx file extension. You can read more in the TypeScript Generics. Discussing naming conventions
```

For 2, this would be the generic types section of the spec, I think: https://github.com/Microsoft/TypeScript/blob/730f18955dc17068be33691f0fb0e0285ebbf9f5/doc/spec.md#3.7

3 might be related to 1

4 is TSX specific for sure as well
